File: 560_subarray_sum_equals_k.rb
=begin
Given an array of integers and an integer k, you need to find the total number of continuous subarrays whose sum equals to k.

Example 1:
Input:nums = [1,1,1], k = 2
Output: 2
Note:
The length of the array is in range [1, 20,000].
The range of numbers in the array is [-1000, 1000] and the range of the integer k is [-1e7, 1e7].
=end

# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def subarray_sum(nums, k)
    k_count = 0
    sum = 0
    count_by_sum = {}
    count_by_sum[0] = 1
    nums.each do |num|
        sum += num
        if count_by_sum[sum - k]
            k_count += count_by_sum[sum - k]
        end
        count_by_sum[sum] ||= 0
        count_by_sum[sum] += 1
    end
    
    return k_count
end

File: problems/560_subarray_sum_equals_k.rs
/*
Given an array of integers nums and an integer k, return the total number of continuous subarrays whose sum equals to k.

 

Example 1:

Input: nums = [1,1,1], k = 2
Output: 2
Example 2:

Input: nums = [1,2,3], k = 3
Output: 2
 

Constraints:

1 <= nums.length <= 2 * 104
-1000 <= nums[i] <= 1000
-107 <= k <= 107
*/

use std::collections::HashMap;

impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut k_count = 0;
        let mut sum = 0;
        let mut map: HashMap<i64, i32> = HashMap::new(); 
        map.insert(0, 1);
        for n in nums {
            sum += n as i64; 
            k_count += map.get(&(sum - k as i64)).unwrap_or(&0);
            *map.entry(sum).or_insert(0) += 1;
        }
        k_count
    }
}

